[FLINK-22483][flink-runtime] Remove CompletedCheckpointStore#recover(…

dmvk opened the pull request:
## What is the purpose of the change

Decoupling recovery of CompletedCheckpointStore from execution graph. We only need to recover it once, when JobMasterRunner gains leadership.
## Brief change log

- Removing `CompletedCheckpointStore#recover()` method
- Replacing `CheckpointRecoveryFactory#createCompletedCheckpointStore` with `CheckpointRecoveryFactory#createRecoveredCompletedCheckpointStore`
- 
- ## Verifying this change

* All existing tests should be passing.

## Does this pull request potentially affect one of the following parts:

  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes)

## Documentation
  - It's a breaking change, so we should mention this in the release notes.

dmvk commented on flink-tests/src/test/java/org/apache/flink/test/checkpointing/CheckpointStoreITCase.java:
```diff
@@ -1,213 +0,0 @@
- * Test that failure on recovery leads to job restart if configured, so that transient recovery
- * failures can are mitigated.
- */
-public class CheckpointStoreITCase extends TestLogger {
```
This test is no longer valid as we don't recover checkpoint store during local recovery.

dmvk commented:
CI test failure -> https://issues.apache.org/jira/browse/FLINK-23590

dmvk commented:
@flinkbot run azure

dmvk commented:
@flinkbot run azure

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedHaServicesWithLeadershipControl.java:
```diff
@@ -36,7 +36,15 @@ public EmbeddedHaServicesWithLeadershipControl(Executor executor) {
-                        n -> new EmbeddedCompletedCheckpointStore(),
+                        (maxCheckpoints, previous) -> {
+                            if (previous != null) {
+                                final EmbeddedCompletedCheckpointStore cast =
+                                        (EmbeddedCompletedCheckpointStore) previous;
```
I am wondering whether we can make `PerJobCheckpointReocveryFactory` type safe by introducing a type parameter.

dmvk replied:
👍 fixed

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorRestoringTest.java:
```diff
@@ -208,7 +211,12 @@ public void testRestoreLatestCheckpointedState() throws Exception {
         tasks.add(jobVertex2);
 
-        assertTrue(coord.restoreLatestCheckpointedStateToAll(tasks, false));
+        // set up the coordinator and validate the initial state
+        final CompletedCheckpointStore secondStore =
+                new EmbeddedCompletedCheckpointStore(1, store.getSuspendedCheckpoints());
+        final CheckpointCoordinator secondCoordinator =
+                coordinatorBuilder.setCompletedCheckpointStore(secondStore).build();
+        assertTrue(secondCoordinator.restoreLatestCheckpointedStateToAll(tasks, false));
```
With the change of this commit, the test case could actually be split into two parts:

1) Successful checkpoints are stored in the `CompletedCheckpointStore`.
2) If the `CompletedCheckpointStore` contains a checkpoint, then it is used by the `CheckpointCoordinator` to restore state.

1) might already be covered by some other tests.

dmvk replied:
refactored & split in two "stages"

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:
```diff
@@ -2792,24 +2792,23 @@ public void testSharedStateRegistrationOnRestore() throws Exception {
 
         // set up the coordinator and validate the initial state
-        CheckpointCoordinator checkpointCoordinator =
+        final CheckpointCoordinatorBuilder coordinatorBuilder =
```
The same applies to this test. If we can create the proper checkpoint (incremental ones), then we don't need to create the checkpoint via running the `ExecutionGraph`.

dmvk replied:
I agree with a sentiment of this, but it seems that creating checkpoints via execution graph is actually the easiest approach.

I did some research and it would be probably possible to create a wrapper around PendingCheckpoint (which is hard to setup).

I'd rather leave this one as is for now and maybe create a follow-up issue for refactoring this if you want. This would probably take me another day as it would require deep diving into internal data structures of checkpoints.

tillrohrmann replied:
Ok, this makes sense.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/DefaultCompletedCheckpointStoreUtilsTest.java:
```diff
@@ -0,0 +1,137 @@
+import static org.junit.Assert.assertThrows;
+
+/** Tests related to {@link DefaultCompletedCheckpointStore}. */
+public class DefaultCompletedCheckpointStoreUtilsTest {
```
```suggestion
public class DefaultCompletedCheckpointStoreUtilsTest extends TestLogger {
```

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/DefaultCompletedCheckpointStoreUtilsTest.java:
```diff
@@ -0,0 +1,137 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThrows;
+
+/** Tests related to {@link DefaultCompletedCheckpointStore}. */
```
```suggestion
/** Tests related to {@link DefaultCompletedCheckpointStoreUtils}. */
```

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreTest.java:
```diff
@@ -180,17 +89,19 @@ private void testDownloadInternal(
+            DefaultCompletedCheckpointStoreUtils.retrieveCompletedCheckpoints(
+                    checkpointsInZooKeeper, zooKeeperCheckpointStoreUtil);
+        } catch (Exception exception) {
+            findThrowable(exception, ExpectedTestException.class)
+                    .ifPresent(ExceptionUtils::rethrow);
+            rethrow(exception);
```
You remove the `expected = ` and use `FlinkMatchers.containsCause` instead. This might be a bit simpler.

dmvk replied:
fixed

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/checkpointing/CheckpointStoreITCase.java:
```diff
@@ -63,7 +61,10 @@
+                    .set(
+                            HighAvailabilityOptions.HA_MODE,
+                            BlockingHighAvailabilityServiceFactory.class.getName())
+                    .set(JobManagerOptions.SCHEDULER, Adaptive);
```
Why did you set the `Adaptive` scheduler here?

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/checkpointing/CheckpointStoreITCase.java:
```diff
@@ -126,68 +147,39 @@ public HighAvailabilityServices createHAServices(
 
                 @Override
                 public CheckpointRecoveryFactory getCheckpointRecoveryFactory() {
```
Would it work to instantiate a `TestingHighAvailabilityServices` with a special `CheckpointRecoveryFactory` instead of creating a sub class of `EmbeddedHaServices`?

dmvk replied:
Not without implementing proper cleanup of leader election services for `TestingHighAvailabilityServices` (otherwise I'll end up with something like `Caused by: java.lang.IllegalStateException: leader retrieval service is already started`).

I think implementing this would basically lead to re-implemention of `EmbeddedHaServices` for purpose of this test.

WDYT?

dmvk replied:
There are actually more sub-classes of EmbeddedHaServices serving the same / similar purpose. All are intended to be used with MiniCluster, probably for the same reason I've mentioned above

dmvk replied:
I was able to get rid of all them, by adding second constructor to `EmbeddedHaServicesWithLeadershipControl`, that allows overriding of `CheckpointRecoveryFactory`. 🎉

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @dmvk. The overall change looks good. I had a few comments on the new contract of the `CompletedCheckpointStore` and whether we can simplify the `PerJobCheckpointRecoveryFactory`. PTAL.

edu05 commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/EmbeddedCompletedCheckpointStore.java:
```diff
@@ -35,23 +36,24 @@
-        Preconditions.checkArgument(maxRetainedCheckpoints > 0);
-        this.maxRetainedCheckpoints = maxRetainedCheckpoints;
+    public EmbeddedCompletedCheckpointStore(int maxRetainedCheckpoints) {
+        this(maxRetainedCheckpoints, Collections.emptyList());
```
why remove the precondition?

edu05 commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DefaultCompletedCheckpointStore.java:
```diff
@@ -95,18 +86,14 @@ public DefaultCompletedCheckpointStore(
-
         this.checkpointStateHandleStore = checkNotNull(stateHandleStore);
-
         this.completedCheckpoints = new ArrayDeque<>(maxNumberOfCheckpointsToRetain + 1);
```
not necessary a comment related to this PR's changes but `maxNumberOfCheckpointsToRetain` does not enforce `completedCheckpoints` from growing. is that something that needs to be addressed?

dmvk replied:
Also we can not enforce this when constructing a new CompletedCheckpointStore, because we've not yet registered completed checkpoints with a SharedStateRegistry, so this might break incremental checkpoints.

There is actually no change from the previous behavior.

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DefaultCompletedCheckpointStore.java:
```diff
@@ -95,18 +86,14 @@ public DefaultCompletedCheckpointStore(
-
         this.completedCheckpoints = new ArrayDeque<>(maxNumberOfCheckpointsToRetain + 1);
-
+        this.completedCheckpoints.addAll(completedCheckpoints);
```
That's a good idea. I'll add a safeguard to `DefaultCompletedCheckpointStore#addCheckpoint`, that should be only method that can mutate its state.

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PerJobCheckpointRecoveryFactory.java:
```diff
@@ -61,6 +65,7 @@ public CheckpointIDCounter createCheckpointIDCounter(JobID jobId) {
     public static CheckpointRecoveryFactory useSameServicesForAllJobs(
             CompletedCheckpointStore store, CheckpointIDCounter counter) {
-        return new PerJobCheckpointRecoveryFactory(n -> store, () -> counter);
+        return new PerJobCheckpointRecoveryFactory(
+                (maxCheckpoints, previous) -> store, () -> counter);
```
I've managed to completely remove this. For tests that are not using EmbbededCCS, I've added a new factory method to simplify the usage a little bit. This seems to cover all of the current usages.

```java
    @VisibleForTesting
    public static <T extends CompletedCheckpointStore>
    CheckpointRecoveryFactory withoutCheckpointStoreRecovery(
            Function<Integer, T> storeFn) {
        return new PerJobCheckpointRecoveryFactory<>(
                (maxCheckpoints, previous) -> {
                    if (previous == null) {
                        throw new UnsupportedOperationException("Checkpoint store recovery is not supported.");
                    }
                    return storeFn.apply(maxCheckpoints);
                });
    }
```

tillrohrmann replied:
👍

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/EmbeddedCompletedCheckpointStore.java:
```diff
@@ -35,23 +36,24 @@
     private final ArrayDeque<CompletedCheckpoint> checkpoints = new ArrayDeque<>(2);
 
-    private final Collection<CompletedCheckpoint> suspended = new ArrayDeque<>(2);
+    private final Collection<CompletedCheckpoint> suspended =
+            Collections.synchronizedList(new ArrayList<>());
```
This is still needed as we use it later for constructing a new checkpoint store after failovaer (this simulates persisting these checkpoints in some kind of remote storage after shutting down store with non-globally terminal state.

It's also good that we can only recover these checkpoints after shutdown has been called.

tillrohrmann replied:
Fair enough. Sounds good.

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpointStore.java:
```diff
@@ -32,14 +32,6 @@
 
     Logger LOG = LoggerFactory.getLogger(CompletedCheckpointStore.class);
 
-    /**
-     * Recover available {@link CompletedCheckpoint} instances.
-     *
-     * <p>After a call to this method, {@link #getLatestCheckpoint(boolean)} returns the latest
-     * available checkpoint.
-     */
-    void recover() throws Exception;
```
added

dmvk commented:
@tillrohrmann Thanks for the review. I've addressed most of your comments, this led to a nice refactoring. There are still some discussions open though, PTAL.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PerJobCheckpointRecoveryFactory.java:
```diff
@@ -21,46 +21,63 @@
+
+    @VisibleForTesting
+    public static <T extends CompletedCheckpointStore>
+            CheckpointRecoveryFactory withoutCheckpointStoreRecovery(Function<Integer, T> storeFn) {
```
```suggestion
            CheckpointRecoveryFactory withoutCheckpointStoreRecovery(IntFunction<T> storeFn) {
```

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:
```diff
@@ -68,11 +67,20 @@
+                new PerJobCheckpointRecoveryFactory<EmbeddedCompletedCheckpointStore>(
+                        (maxCheckpoints, previous) -> {
+                            if (previous != null) {
+                                // First job attempt failed before cleaning up the checkpoint store.
+                                // This means we need to use the actual checkpoint store state for
+                                // recovery (instead of suspended checkpoints as usual).
```
Hmm, if we remove the `suspendedCheckpoints` field from the `EmbeddedCompletedCheckpointStore` and remove checkpoints from `checkpoints` on `shutDown(GloballyTerminalState)`, then we wouldn't need this extra logic here. The contract would then be: Whatever is returned by `store.getAllCheckpoints()` will be recovered.

dmvk replied:
I think we should also expose shutdown status for some advanced assert (like this one), but this should be doable 🤔

dmvk replied:
fixed

tillrohrmann reviewed (approved):
Thanks for updating this PR @dmvk. The changes look good to me. I had a few minor comments left. PTAL. I think we can merge this PR once the comments have been resolved.

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DefaultCompletedCheckpointStoreUtils.java:
```diff
@@ -0,0 +1,116 @@
+/** Helper methods related to {@link DefaultCompletedCheckpointStore}. */
+public class DefaultCompletedCheckpointStoreUtils {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LoggerFactory.class);
```
🤦

dmvk commented:
Fails with https://issues.apache.org/jira/browse/FLINK-20410

dmvk commented:
@flinkbot run azure
